File: reef-io/src/main/java/com/microsoft/reef/io/network/nggroup/impl/Sender.java
/**
 * Copyright (C) 2014 Microsoft Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.microsoft.reef.io.network.nggroup.impl;

import com.microsoft.reef.exception.evaluator.NetworkException;
import com.microsoft.reef.io.network.Connection;
import com.microsoft.reef.io.network.group.operators.AbstractGroupCommOperator;
import com.microsoft.reef.io.network.impl.NetworkService;
import com.microsoft.reef.io.network.proto.ReefNetworkGroupCommProtos.GroupCommMessage;
import com.microsoft.reef.io.network.util.StringIdentifierFactory;
import com.microsoft.wake.Identifier;
import com.microsoft.wake.IdentifierFactory;

import java.util.logging.Logger;

/**
 *
 */
public class Sender extends AbstractGroupCommOperator {

  private static final Logger LOG = Logger.getLogger(Sender.class.getName());

  private final NetworkService<GroupCommMessage> netService;
  private final IdentifierFactory idFac = new StringIdentifierFactory();

  public Sender(final NetworkService<GroupCommMessage> netService) {
    this.netService = netService;
  }

  public void send(final GroupCommMessage msg) throws NetworkException {
    final String dest = msg.getDestid();
    send(msg, dest);
  }

  public void send(final GroupCommMessage msg, final String dest) throws NetworkException {
    final Identifier destId = idFac.getNewInstance(dest);
    final Connection<GroupCommMessage> link = netService.newConnection(destId);
    link.open();
    link.write(msg);
  }
}

File: lang/cs/Org.Apache.REEF.Network/Group/Operators/Impl/Sender.cs
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

using System;
using Org.Apache.REEF.Network.Group.Driver.Impl;
using Org.Apache.REEF.Network.NetworkService;
using Org.Apache.REEF.Tang.Annotations;
using Org.Apache.REEF.Wake;

namespace Org.Apache.REEF.Network.Group.Operators.Impl
{
    /// <summary>
    /// Group Communication operator used to do point-to-point communication between named Tasks.
    /// It uses Writable classes
    /// </summary>
    internal sealed class Sender
    {
        private readonly INetworkService<GeneralGroupCommunicationMessage> _networkService;
        private readonly IIdentifierFactory _idFactory;

        /// <summary>
        /// Creates a new Sender.
        /// </summary>
        /// <param name="networkService">The network services used to send messages.</param>
        /// <param name="idFactory">Used to create IIdentifier for GroupCommunicationMessages.</param>
        [Inject]
        private Sender(
            StreamingNetworkService<GeneralGroupCommunicationMessage> networkService,
            IIdentifierFactory idFactory)
        {
            _networkService = networkService;
            _idFactory = idFactory;
        }

        /// <summary>
        /// Send the GroupCommunicationMessage to the Task whose name is
        /// included in the message.
        /// </summary>
        /// <param name="message">The message to send.</param>
        internal void Send(GeneralGroupCommunicationMessage message)
        {
            if (message == null)
            {
                throw new ArgumentNullException("message");
            }
            if (string.IsNullOrEmpty(message.Destination))
            {
                throw new ArgumentException("Message destination cannot be null or empty");
            }

            IIdentifier destId = _idFactory.Create(message.Destination);
            var conn = _networkService.NewConnection(destId);
            conn.Open();
            conn.Write(message);
        }
    }
}
